kv/dist_sender.go

```diff
@@ -328,17 +334,80 @@ func (ds *DistSender) getRangeDescriptor(key proto.Key) ([]proto.RangeDescriptor
-// could not be sent. Note that the reply may contain a higher level
-// error and must be checked in addition to the RPC error.
-func (ds *DistSender) sendRPC(desc *proto.RangeDescriptor, method string, args proto.Request, reply proto.Response) error {
+func (ds *DistSender) optimizeReplicaOrder(replicas proto.ReplicaSlice) rpc.OrderingPolicy {
+	// Unless we know better, send the RPCs randomly.
+	order := rpc.OrderingPolicy(rpc.OrderRandom)
+	nodeDesc := ds.getNodeDescriptor()
+	// If we don't know which node we're on, don't optimize anything.
+	if nodeDesc == nil {
+		return order
+	}
+	// Sort replicas by attribute affinity, which we treat as a stand-in for
+	// proximity (for now).
+	if replicas.SortByCommonAttributePrefix(nodeDesc.Attrs.Attrs) > 0 {
+		// There's at least some attribute prefix, and we hope that the
+		// replicas that come early in the slice are now located close to
+		// us and hence better candidates.
+		order = rpc.OrderStable
+	}
+	return order
+}
+
+// getNodeDescriptor returns ds.nodeDescriptor, but makes an attempt to load
+// it from the Gossip network if a nil value is found.
+// This can only succeed after the initial bootstrap is done, but a descriptor
+// will always be available during normal operation.
+func (ds *DistSender) getNodeDescriptor() *storage.NodeDescriptor {
+	if ds.nodeDescriptor != nil {
+		return ds.nodeDescriptor
+	}
+	ownNodeID := ds.gossip.GetNodeID()
+	if nodeDesc, err := ds.gossip.GetInfo(
+		gossip.MakeNodeIDKey(ownNodeID)); err == nil && ownNodeID > 0 {
+		// Rearrange the replicas so that those replicas with long common
+		// prefix of attributes end up first. If there's no prefix, this is a
+		// no-op.
+		ds.nodeDescriptor = nodeDesc.(*storage.NodeDescriptor)
+	} else if ownNodeID > 0 {
+		// A normal node should always have that information set in Gossip when
+		// bootstrapped, so this is worth complaining about.
+		log.Warningf("own NodeDescriptor %d not available via Gossip",
+			ownNodeID)
+	} else {
+		log.V(1).Infof("gossip unaware of NodeID, node likely still bootstrapping")
+	}
+	return ds.nodeDescriptor
+}
+
+// sendRPC sends one or more RPCs to replicas from the supplied proto.Replica
+// slice. First, replicas which have gossiped addresses are corralled (and
+// rearranged depending on proximity and whether the request needs to go to a
+// leader) and then sent via rpc.Send, with requirement that one RPC to a
+// server must succeed. Returns an RPC error if the request could not be sent.
+// Note that the reply may contain a higher level error and must be checked in
+// addition to the RPC error.
+func (ds *DistSender) sendRPC(desc *proto.RangeDescriptor, method string,
+	args proto.Request, reply proto.Response) error {
 	if len(desc.Replicas) == 0 {
 		return util.Errorf("%s: replicas set is empty", method)
 	}
 
+	// Rearrange the replicas suitably, and return the desired order.
+	order := ds.optimizeReplicaOrder(proto.ReplicaSlice(desc.Replicas))
```

Comment:
Might be better to do this re-ordering on the cached descriptor so that it doesn't have to be done on every RPC. Oh, is `desc` the cached descriptor? This is going to be problematic as multiple goroutines can be re-ordering the replicas concurrently.